### PR TITLE
Added TrainOptions class and fixed some comments

### LINKED ISSUE #135: Make a `TrainOptions` class
This will allow not passing many parameters (which seems to grow with every PR lol) to the training function.

### PR DESCRIPTION
**Pull request by fshcat**
(resolves #135 )

### CODE REVIEW

**Comment by nashirj** [comment]
We discussed this and decided to wait for the checkpointing branch to be closed so that parameters related to model saving can be included in the TrainOptions class.

**Review by ryPattillo — COMMENTED** [review]
Looks like a good start, there are a bunch of more training parameters that we need to consider but that can be added on another pull request

File: chess-AI/train.py [hunk 1]
```diff
@@ -16,6 +13,33 @@
...
+            learning_rate: the learning rate for the optimizer
+            momentum: additional parameter for the optimizer
+            weight_decay: loss term encouraging smaller weights
+            epochs: number of epochs
+            batch_size: the batch size for SGD
+            games: number of games to run when creating dataset
+            device: the device being used to train (either CPU or GPU)
+            save_path: path for model checkpointing
+            num_saved_models: number of models to store
+        """
```
Comment (nashirj) [inline chess-AI/train.py:29]: This entire block should be indented one level less, so that the start and end """ line up

File: chess-AI/train.py [hunk 2]
```diff
@@ -118,20 +142,13 @@ def create_dataset(games, move_approximator, val_approximator=None):
...
         nnet: the neural network
-        learning_rate: the learning rate for the optimizer
-        momentum: additional parameter for the optimizer
-        weight_decay: loss term encouraging smaller weights
-        epochs: number of epochs
-        batch_size: the batch size for SGD
-        device: the device being used to train (either CPU or GPU)
-        save_path: path for model checkpointing
-        num_saved_models: number of models to store
+        options: settings/hyperparameters for training
```
Comment (nashirj) [inline chess-AI/train.py:151]: Can you pls specify that the type is TrainOptions so people know where to look to see all the available hyperparameters

File: chess-AI/train.py [hunk 3]
```diff
@@ -247,25 +231,56 @@ def main():
...
+    stockfish_options = TrainOptions(learning_rate=0.1, momentum=0.9, weight_decay=0.001, epochs=10, batch_size=8, games=1000,
+                                     device=device, save_path=model_path, num_saved_models=num_saved_models, overwrite=overwrite)
+
+    train_on_stockfish(nnet, elo=1000, depth=3, dataset_path=dataset_path, options=stockfish_options)
+
+    # Train network using MCTS
+    mcts_options = TrainOptions(learning_rate=0.1, momentum=0.9, weight_decay=0.001, epochs=100, batch_size=8, games=100,
+                                device=device, save_path=model_path, num_saved_models=num_saved_models, overwrite=overwrite)
+
+    train_on_mcts(nnet, exploration=5, training_episodes=5, mcts_simulations=100, options=mcts_options)
```
Comment (nashirj) [inline chess-AI/train.py:283]: Can you pls make variables for all the parameters specified in the function call? It'll make it easier for @ryPattillo and I to update this code to handle either the dashboard/reading from json. You can initialize the variables with the values here, and we can use those for reference as the default for the hyperparams.json file we'll make

**Review by nashirj — CHANGES_REQUESTED** [review]
This looks much cleaner, nice work and thanks for doing this.

Couple things in addition to my other comments, I tried running the code and got this stack trace:
```bash
Traceback (most recent call last):
  File "train.py", line 287, in <module>
    main()
  File "train.py", line 283, in main
    train_on_mcts(nnet, exploration=5, training_episodes=2, mcts_simulations=2, options=mcts_options)
  File "train.py", line 251, in train_on_mcts
    dataset = create_dataset(options.games, mcts_moves)
  File "train.py", line 131, in create_dataset
    game_data = [training_game(val_approximator, move_approximator) for _ in range(games)]
  File "train.py", line 131, in <listcomp>
    game_data = [training_game(val_approximator, move_approximator) for _ in range(games)]
  File "train.py", line 72, in training_game
    moves, move_probs, move = move_approximator(board)
  File "train.py", line 249, in <lambda>
    mcts_moves = lambda board: mcts.get_tree_results(mcts_simulations, nnet, board, temperature=5)
  File "/home/nashirj/Documents/school/UCF/clubs/knightros-gambit/chess-AI/mcts.py", line 57, in get_tree_results
    self.search(board, nnet)
TypeError: search() missing 1 required positional argument: 'device'
```

Also, prob not in the scope of this PR, but there was also a warning about creating a tensor from a list of arrays being slow:
```
train.py:136: UserWarning: Creating a tensor from a list of numpy.ndarrays is extremely slow. Please consider converting the list to a single numpy.ndarray with numpy.array() before converting to a tensor. (Triggered internally at  /opt/conda/conda-bld/pytorch_1640811806235/work/torch/csrc/utils/tensor_new.cpp:201.)
  move_probs = torch.tensor([move_prob for game in game_data for move_prob in game[2]]).float()
```
prob not hard to fix that warning while you're doing other stuff if you don't mind.

**Review by nashirj — APPROVED** [review]

**Pull request merged**